Added Crypto Policy rule for JRE.

matejak opened the pull request:
The rule ensures that the service adheres to the system-wide crypto policy.

Remarks:
* The rule is part of the Fedora `standard` profile.
* The SSG tests pass, I recommend using the Docker backend.
* The Ansible fix passes the SSG Test Suite if you make sure that Python is installed in the backend.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/f13ee258-6824-44fb-a5b7-a480ff392b44) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/f13ee258-6824-44fb-a5b7-a480ff392b44) completed: **No new issues**

shawndwells commented:
Should this be added to the JRE content, vs Fedora?

https://github.com/ComplianceAsCode/content/tree/master/jre

redhatrises commented:
Is this not configurable in `/etc/.java`?

matejak commented:
@redhatrises I have found no indication that the `java.security` file could be located elsewhere than the `lib\security` subdirectory of the respective JRE.

@shawndwells This is a tough question - there will be more rules that relate to crypto policies, so I decided to go this way in #3579. You typically aren't interested in SSH server to satisfy all requirements, ignoring other services that use encryption. This group aggregates rules that are all about encryption, and they are likely to fail for the same reason. So instead of getting five FAILs across multiple groups, you will get them in one group, which points you to the root cause.

redhatrises commented:
> @redhatrises I have found no indication that the `java.security` file could be located elsewhere than the `lib\security` subdirectory of the respective JRE.
> 
> @shawndwells This is a tough question - there will be more rules that relate to crypto policies, so I decided to go this way in #3579. You typically aren't interested in SSH server to satisfy all requirements, ignoring other services that use encryption. This group aggregates rules that are all about encryption, and they are likely to fail for the same reason. So instead of getting five FAILs across multiple groups, you will get them in one group, which points you to the root cause.

Agreed that this is a tough question. However after mulling it over, the rule and associated content should be in the JRE content and not the OS content. A couple of reasons for this:

1. JRE is an application that does not have to be installed on Fedora. So, you would get a fail if JRE is not installed (which there are scenarios where this is the case) which also means it should not be enabled by default in the `standard` profile.
2. If JRE is installed, setting the crypto policy to use the system crypto policy is not enough for securing Java. More has to be done which is what the JRE content is for.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c9490ea5-4601-483f-80ca-26625cdddaa4) was created.

matejak commented:
@shawndwells , @redhatrises I have moved the rule to the jre section and renamed it, so it conforms to other Java rules. The downside of this is that it can't be easily tested by the test suite, but tests are passing for the earlier commit, so it may not be an issue.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/c9490ea5-4601-483f-80ca-26625cdddaa4) completed: **No new issues**

yuumasato commented on jre/guide/java/java_jre_configure_crypto_policy/ansible/shared.yml:
```diff
@@ -1,4 +1,4 @@
-# platform = multi_platform_all
+# platform = multi_platform_all,Java Runtime Environment
```
Shouldn't the platform be only `Java Runtime Environment`? 
Do you keep it for testing reasons?
I have same comment for Bash and OVAL content.

matejak replied:
As the remediation content won't get anywhere without the respective rule, I don't think that it makes sense to restrict it in that way. The Ansible playbook will do its job on Fedora, and maybe on some other Linux distribution, so let's not restrict Ansible for the Java "platform".

yuumasato commented:
@redhatrises What is the practice in the field regarding JRE content?
When people are concerned about Java in their systems, do they scan the first the OS, for example Fedora or RHEL content, and then scan the Java configurations, using JRE content?

yuumasato commented on tests/data/group_system/group_crypto_policies/rule_configure_jre_crypto_policy/commented_out.fail.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# profiles = xccdf_org.ssgproject.content_profile_standard
```
There is no `standard` profile for JRE. Should there be one?
Because the only profile for JRE is one derived from STIG.

yuumasato replied:
I'm okay to leave the rule out of any profile for now. Adding a new profile is totally out of scope for this PR.

matejak replied:
I have marked this for the `(all)` profile, that was introduced in the recent OpenSCAP.

yuumasato commented on jre/guide/java/java_jre_configure_crypto_policy/oval/shared.xml:
```diff
@@ -0,0 +1,30 @@
+    <metadata>
+      <title>Configure JRE to use System Crypto Policy.</title>
+      <affected family="unix">
+        <platform>multi_platform_fedora</platform>
```
I think there is no need to keep platform Fedora.
Platform `Java Runtime Environment` is applicable as long as java is installed.

matejak commented:
After discussing the meaning of platform, I have removed other than Java-related platforms from the content.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/21a42cdb-c49a-46bf-9270-bb6dfb0048b7) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/21a42cdb-c49a-46bf-9270-bb6dfb0048b7) completed: **No new issues**

jan-cerny commented:
I cannot run the tests.
```
[jcerny@thinkpad tests{crypto_jre}]$ python3 test_suite.py rule --debug --libvirt qemu:///system ssg-test-suite-fedora --datastream ../build/ssg-jre-ds.xml --xccdf-id scap_org.open-scap_cref_ssg-jre-xccdf-1.2.xml java_jre_configure_crypto_policy
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/scap-security-guide/tests/logs/rule-custom-2018-11-12-1410/test_suite.log                                                                                                        
ERROR - No matching rule ID found for '['java_jre_configure_crypto_policy']'
ERROR - Nothing has been tested!
```

matejak commented:
The problem boils down to the fact that the rule is not part of any profile. Therefore, it has to be used with the `(all)` profile, which the unpatched testsuite can't handle, and even `oscap-ssh` can't handle it. The problem is that `(all)` causes shell syntax errors.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/3848c8f9-e8df-45f1-83d6-84dc4dec0c17) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/3848c8f9-e8df-45f1-83d6-84dc4dec0c17) completed: **No new issues**

redhatrises commented on linux_os/guide/system/crypto_policies/group.yml:
```diff
@@ -0,0 +1,5 @@
+documentation_complete: true
```
@matejak this file needs to be removed from this PR, but after that, I think that it is good to merge.

jan-cerny commented on jre/guide/java/java_jre_configure_crypto_policy/rule.yml:
```diff
@@ -0,0 +1,27 @@
+    the <pre>java.security</pre> doesn't contain the <pre>security.useSystemPropertiesFile=true</pre> line
+
+ocil: |-
+    To verify that Java uses the system crypro policy, check out that the respective JRE config file
```
@matejak crypro

jan-cerny commented:
@matejak please remove documentation_complete and change "crypro", and then it can be merged, it works great on my machine

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/65af6007-602e-48cd-a25c-c320d737f14c) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/ComplianceAsCode/content/inspections/65af6007-602e-48cd-a25c-c320d737f14c) completed: **No new issues**

matejak commented:
Thank you all for looking into this PR, there is no crypro and no stray group YAML any more.

redhatrises commented:
LGTM
